route-rs-runtime/src/link/async_link.rs

```diff
@@ -1,43 +1,98 @@
-use crate::element::AsyncElement;
+use crate::element::Element;
 use crate::link::task_park::*;
-use crate::link::PacketStream;
+use crate::link::{ElementLinkBuilder, Link, LinkBuilder, PacketStream};
 use crossbeam::atomic::AtomicCell;
 use crossbeam::crossbeam_channel;
 use crossbeam::crossbeam_channel::{Receiver, Sender, TryRecvError};
 use futures::{Async, Future, Poll, Stream};
 use std::sync::Arc;
 
-/// The AsyncLink is a wrapper to create and contain both sides of the
-/// link, the Ingressor, which intakes and processes packets, and the Egressor,
-/// which provides an interface where the next element retrieves the output
-/// packet.
-pub struct AsyncLink<E: AsyncElement> {
-    pub ingressor: AsyncIngressor<E>,
-    pub egressor: AsyncEgressor<E::Output>,
+#[derive(Default)]
+pub struct AsyncLinkBuilder<E: Element> {
+    in_stream: Option<PacketStream<E::Input>>,
+    element: Option<E>,
+    queue_capacity: usize,
 }
 
-impl<E: AsyncElement> AsyncLink<E> {
-    pub fn new(input_stream: PacketStream<E::Input>, element: E, queue_capacity: usize) -> Self {
+impl<E: Element> AsyncLinkBuilder<E> {
+    pub fn new() -> Self {
+        AsyncLinkBuilder {
+            in_stream: None,
+            element: None,
+            queue_capacity: 10,
+        }
+    }
+
+    pub fn ingressor(self, in_stream: PacketStream<E::Input>) -> Self {
+        AsyncLinkBuilder {
+            in_stream: Some(in_stream),
+            element: self.element,
+            queue_capacity: self.queue_capacity,
+        }
+    }
+
+    /// Changes queue_capacity, default value is 10.
+    /// Valid range is 1..=1000
+    pub fn queue_capacity(self, queue_capacity: usize) -> Self {
         assert!(
             queue_capacity <= 1000,
             format!("Async Element queue_capacity: {} > 1000", queue_capacity)
         );
         assert_ne!(queue_capacity, 0, "queue capacity must be non-zero");
 
-        let (to_egressor, from_ingressor) =
-            crossbeam_channel::bounded::<Option<E::Output>>(queue_capacity);
-        let task_park: Arc<AtomicCell<TaskParkState>> =
-            Arc::new(AtomicCell::new(TaskParkState::Empty));
+        AsyncLinkBuilder {
+            in_stream: self.in_stream,
+            element: self.element,
+            queue_capacity,
+        }
+    }
+}
+
+impl<E: Element + Send + 'static> LinkBuilder<E::Input, E::Output> for AsyncLinkBuilder<E> {
+    fn ingressors(self, mut in_streams: Vec<PacketStream<E::Input>>) -> Self {
+        assert_eq!(
+            in_streams.len(),
+            1,
+            "AsyncLink may only take 1 input stream"
```

Comment:
I may answer this later as I keep reading, but just confirming that we're keeping `Async` as 1 - 1, where we would use Link composition with `Join` and `Tee` in order to get the M - N effect?

Reply:
Exactly, which is why I changed it to queue link instead of async link